### PR TITLE
Prevent `GetFunctionCodeSigningConfig` calls when a functions packageType is `Image`

### PR DESCRIPTION
**Pull request by A-Hilaly**
Issue #, if available: https://github.com/aws-controllers-k8s/community/issues/1117

Description of changes:
Functions code signing config should only be called when a function is
created using an s3bucket and a key. Functions created using a container
image cannot get a code signing configuration.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/resource/function/hooks.go [hunk 1]
```diff
@@ -500,17 +500,19 @@ func (rm *resourceManager) setResourceAdditionalFields(
...
-	getFunctionCodeSigningConfigOutput, err = rm.sdkapi.GetFunctionCodeSigningConfigWithContext(
-		ctx,
-		&svcsdk.GetFunctionCodeSigningConfigInput{
-			FunctionName: ko.Spec.Name,
-		},
-	)
-	rm.metrics.RecordAPICall("GET", "GetFunctionCodeSigningConfig", err)
-	if err != nil {
-		return err
+	if ko.Spec.PackageType != nil && *ko.Spec.PackageType != "Image" {
```
Comment (vijtrip2) [inline pkg/resource/function/hooks.go:503]: nit:
```suggestion
	if ko.Spec.PackageType != nil && *ko.Spec.PackageType == "Zip" {
```

**Review by vijtrip2 — APPROVED** [review]

File: pkg/resource/function/hooks.go [hunk 2]
```diff
@@ -500,17 +500,19 @@ func (rm *resourceManager) setResourceAdditionalFields(
...
+			&svcsdk.GetFunctionCodeSigningConfigInput{
+				FunctionName: ko.Spec.Name,
+			},
+		)
+		rm.metrics.RecordAPICall("GET", "GetFunctionCodeSigningConfig", err)
+		if err != nil {
+			return err
+		}
+		ko.Spec.CodeSigningConfigARN = getFunctionCodeSigningConfigOutput.CodeSigningConfigArn
 	}
```
Comment (jaypipes) [inline pkg/resource/function/hooks.go:516]: Consider adding a condition of type [`ACK.Advisory`](https://github.com/aws-controllers-k8s/runtime/blob/d2f3189f3d5b1af4c7093397fef6d8eeb1edc992/apis/core/v1alpha1/conditions.go#L47) in an else block here when the user specifies a code signing configuration for functions with package type != "Zip"...

That would provide the user with some helpful information indicating their desired spec is incorrect.

**Review by jaypipes — CHANGES_REQUESTED** [review]
I'd like to see an e2e test added that stresses this code path please! :)

**Comment by A-Hilaly** [comment]
>@A-Hilaly where do we stand with this PR? :)

This PR will need more work than expected. We will need proper support for Images based lambda functions (Initializers, custom hooks and proper e2e for it).
Updated #1062

**Comment by gagan-eg** [comment]
Do you know the timeline for this to be merged? We switched to using image based Function because currently S3 Key change updates are not supported.

**Comment by A-Hilaly** [comment]
> Do you know the timeline for this to be merged? We switched to using image based Function because currently S3 Key change updates are not supported.

@gagan-eg very soon :)

File: pkg/resource/function/hooks.go [hunk 3]
```diff
@@ -500,17 +503,27 @@ func (rm *resourceManager) setResourceAdditionalFields(
...
+				FunctionName: ko.Spec.Name,
+			},
+		)
+		rm.metrics.RecordAPICall("GET", "GetFunctionCodeSigningConfig", err)
+		if err != nil {
+			return err
+		}
+		ko.Spec.CodeSigningConfigARN = getFunctionCodeSigningConfigOutput.CodeSigningConfigArn
+	}
+	if ko.Spec.PackageType != nil && *ko.Spec.PackageType == "Image" && ko.Spec.CodeSigningConfigARN != nil {
```
Comment (vijtrip2) [inline pkg/resource/function/hooks.go:520]: should we add an e2e test for this and assert that terminal condition gets set?

These are good scenarios for unit testing but I believe lambda-controller does not have unit testing right now.

**Comment by jaypipes** [comment]
@A-Hilaly looking at the failure in the e2e test, I think you need to add terminal codes to the `generator.yaml` so that things like `InvalidParameterValueException` are understood as terminal exceptions....

**Pull request merged**